### ISSUE TITLE
Instruction fetch can behave incorrectly with delayed grant

### ISSUE THREAD

**Issue by zyedidia** [lowRISC/ibex#1506]
## Observed Behavior
I am observing a bug in which instructions are being fetched/executed that shouldn't be, specifically after commit 72a892d. I used Verilator to run a simulation of Ibex that runs a small test program ([located here](https://github.com/zyedidia/ocibex/blob/ibex_bug/mem/riscvtest.s)) which should ultimately store 25 at memory address 100 and then enter an infinite loop. When Ibex runs the program it stores 7 into memory address 100, before later storing 25 (it should not store 7 first), and when it executes the infinite loop it incorrectly also executes the instruction after the jump, which is an illegal instruction. This causes the core to take an exception and jump to address 0, the exception address for the instantiated core. The start of the program is also located at address 0, which causes it to re-execute the test.

## Expected Behavior
Ibex should only perform one store to address 100: a store of 25, and it should properly jump during the infinite loop without trying to execute an illegal instruction past the end of the program.

## Steps to reproduce the issue
I have uploaded my code to https://github.com/zyedidia/ocibex. The branch `ibex_bug` has Ibex pointing at the buggy commit, while master works properly and points Ibex to the commit just before the buggy one.

To run the example:
```
git clone https://github.com/zyedidia/ocibex
cd ocibex
git checkout ibex_bug
git submodule update --init
make verilator
./obj_dir/test
```

This will print the current instruction for 1000 clock cycles. Note that it keeps repeating the program (due to taking an illegal instruction exception) instead of entering an infinite loop and repeatedly executing `0x0000006f` (`j 0`). This will also generate `trace.vcd` which you can inspect with a waveform viewer. You can run the test on `master` vs. on `ibex_bug` to see the expected vs. observed behavior. The master branch pins an older version of ibex.

## Solution

I believe the problem arises because my connection between the SRAM and Ibex core uses a one-cycle delay before `instr_gnt` is assigned. See the block "SRAM to Ibex" in [soc_top.sv](https://github.com/zyedidia/ocibex/blob/1bf77e1af95a477aa010c614925fe1976f738a86/rtl/soc_top.sv#L111). I copied this from the Arty7 FPGA example, which I assume has a correct connection between SRAM and processor.

I dug into this more, and found that the following line was removed in commit 72a892d, which is suspicious:

```
assign rvalid_or_pmp_err = rdata_outstanding_q[0] & (instr_rvalid_i | rdata_pmp_err_q[0]);
```

The commit then replaces all uses of `rvalid_or_pmp_err` with `instr_valid_i`, but this does not take into account `rdata_outstanding_q[0]`, which may be false and should keep `rvalid` false.

I made [a commit](https://github.com/zyedidia/ibex/commit/c878c92b8d04dbe46be6fc6a24896121eea3644c) to my fork of Ibex that fixes this, and it solves the problem in my testbench. I'm not sure if this is the preferred way to solve the issue, but if desired I can open a pull request.

Thanks for your time.

## My Environment

I'm using Verilator and a top module ([located here](https://github.com/zyedidia/ocibex/blob/ibex_bug/rtl/soc_top.sv)) to simulate a test.

**EDA tool and version:**
Verilator 4.213 devel rev v4.212-55-g5f597dd9

**Operating system:**
Linux

**Version of the Ibex source code:**
I believe the bug is introduced by commit 72a892d.

**Comment by GregAC** [comment]
Thanks for the detailed report @zyedidia 

This is not actually an Ibex bug but a bug in our FPGA example. It's implementing the memory protocol incorrectly (see docs here: https://ibex-core.readthedocs.io/en/latest/03_reference/load_store_unit.html#protocol). It's producing the `instr_gnt` and `instr_rvalid` signals in the same cycle. Instead the `instr_rvalid` must be at least one cycle after `instr_gnt` is asserted.

Before commit https://github.com/lowRISC/ibex/commit/72a892d62ce820d9519825e308df45c05cfbe867 this just so happened to work because of the line you identified. In effect `instr_rvalid` is being ignored if it's not expecting data which makes things work. 

We need that line so `rvalid_or_pmp_err` gets asserted when either valid instruction data is available or when we've requested data but we squashed the request due to a PMP error so the IF stage produces something (valid instruction data or an error indication). Commit https://github.com/lowRISC/ibex/commit/72a892d62ce820d9519825e308df45c05cfbe867  moves the PMP check to between the IF and ID/EX stages so we no longer need to assert on PMP errors, just when we have valid data. In effect `instr_valid_i` now implies `rdata_outstanding_q`.

For correct memory protocol behaviour simply assign req to gnt like so:

```verilog
  assign instr_gnt = instr_req;
  assign data_gnt  = data_req;
```

If I do that in your `soc_top.sv` that appears to fix things.
